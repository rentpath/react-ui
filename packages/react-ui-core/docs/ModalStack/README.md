# Using ModalStack

[ModalStack](https://github.com/rentpath/react-ui/blob/master/packages/react-ui-core/src/ModalStack/ModalStack.js) is a generic component designed to empower developers to add modals to our sites easily and consistently. ModalStack loads modal content dynamically, so it does not add to the page weight until the modal appears

At the heart of **ModalStack** is its render() function:


```javascript
  render() {
    const { currentModal } = this.props

    if (currentModal && currentModal.id && this.modalHost) {
      return createPortal(this.renderModals(), this.modalHost)
    }

    this.toggleBodyClass(false)
    return null
  }
```


__There is no need to render your Modal into the DOM.__ ModalStack uses the React [createPortal](https://reactjs.org/docs/react-dom.html#createportal) function to render children into a specific DOM node (`modalHost)`. The modalHost node has the default id **region-modal**; it is created on demand and appended to the document body.[^1] When a Modal is opened and closed, it is appended to and removed from the `modalHost` node. The currentModal prop is used to keep track of the currently open Modal.

When a Modal is open, ModalStack calls `toggleBodyClass` to add an “`Overlay-lock`” class to the document body. Calling `toggleBodyClass(false) `removes the “`Overlay-lock`” class. This can be used to hide or show other elements, e.g. when the LocationModal is open, we hide the location search results in the header.

![dom](images/dom.png)


## Steps to adding a new Modal:


### 1. Add a new entry to the config object

In rent-js this lives in src/ui/large|small/config/modalConfig.js. In AG, it’s in src/apps/large|small/modalConfig.js. Rent’s modal config looks like so:

```javascript
export const modalDefinitions = {
  [LEAD_MODAL_ID]: {
    name: 'LeadModal',
    resolve: () => import('ui/large/components/LeadModal'),
    overlay: true,
  },
  [THANK_YOU_MODAL_ID]: {
    name: 'ThankYouModal',
    resolve: () => import('ui/large/components/ThankYouModal'),
    overlay: true,
  },
// ...
```


The config keys are unique strings defined in a constants file. The **name** is used to automatically generate webpack chunks for each modal, so each modal’s code is lazily-loaded. The **resolve** key contains a function that loads the modal by **import**ing the corresponding component, and **overlay** controls whether the modal is wrapped by an overlay div or not. The overlay provides a background layer with a click handler that closes the modal.

All of the modals in the config are passed to a **ModalStack** instance in the Layout of the application/view. For example, in rent-js’ large Layout.js, we have:


```javascript
import { modalDefinitions } from 'ui/large/config/modalConfig'

//...  

return (
    <div className={combinedClassOuter}>
      <ModalStack modalDefinitions={modalDefinitions} />
      {HeaderComponent && <HeaderComponent />}
      <section className={theme.Layout_Container}>
        {children}
      </section>
    </div>
  )
```



### 2. Define open and close functions (rent.js only)

In rent.js these live in src/app/store/shared/actions/modal.js. Each modal has a set of named functions, e.g. openSignInModal, closeSignInModal, openFiltersModal, closeFiltersModal, etc. These actions just pass the modal ID to openModal and closeModal. In ag.js, we just use the openModal and closeModal functions and pass in the desired modal ID.

These actions trigger changes to the “modal” (or “modalStack” in the case of ag) slice of the redux state, modifying the openModals list:


![redux](images/redux.png)


In both rent.js and ag.js, ModalStack is connected to redux using the following container:


```javascript
import { connect } from 'react-redux'
import { currentModal as currentModalSelector } from 'app/selectors/modal'
import { closeModal } from 'app/store/shared/actions'
import { ModalStack } from '@rentpath/react-ui-core'

const mapStateToProps = state => ({
  currentModal: currentModalSelector(state),
})

const mapDispatchToProps = {
  onClose: closeModal,
}

export default connect(mapStateToProps, mapDispatchToProps)(ModalStack)
```


Note that even though we’re passing in the `closeModal` function to ModalStack here, you still have to pass a `closeModal` function to your Modal.

ModalStack also supports “stacked” modals. When multiple stacked modals are opened, the topmost modal will display. When it is closed, the next modal down the stack will open, and so on until there are no more modals in the stack. Use the `openStackedModal` action in both ag.js and rent.js if you need this functionality.


### 3. Create a wrapper component for your Modal. 

Your component path must match used by the <code>import</code> statement in the <code>resolve</code> attribute of the modal config created in step 1.

The wrapper can be a normal React Component or PureComponent that renders a <code>react-ui-core Modal.</code>

Here is a simple example:


```javascript
import { Modal } from '@rentpath/react-ui-core'
// …

export default class MyModal extends PureComponent {
  render() {
    const {
      onClose,
      isOpen,
    } = this.props

    return (
      <Modal
        className={theme.LocationModal}
        data-tid="location-input-modal"
        CloseButton={{
          children: <Icon className={theme.LocationModal_CloseButton} svgs={{ close }} />
        }}
        onClose={onClose}
        isOpen={isOpen}
      >
      // … your components go here
      </Modal>
    )
  }
```


Let’s examine each of the props being passed to the **Modal** here:

`className`: provides the top-level class used to style the modal

`data-tid`: used by our end-to-end tests to identify the modal. Technically this is not needed because the Modal component has a default data-tid=”modal”.

`CloseButton:` an object or function that generates a close button element. You only need to worry about the display element; the `onClose` function is connected for you. It is your responsibility to provide this data to the Modal; by default, it renders nothing.

`onClose: `You need to pass in the appropriate close action for your Modal, even though ModalStack has a default function.

`isOpen: `This is needed in most cases to make the Modal open. You don’t need to connect it to anything, just include it here.


<!-- Footnotes themselves at the bottom. -->
## Notes

[^1]:
     It's important for the modal to be outside the normal app HTML so it can float above all other content. Refer to [z-index stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context) for more info.