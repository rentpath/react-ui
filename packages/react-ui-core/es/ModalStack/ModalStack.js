import _extends from "@babel/runtime/helpers/extends";
import _regeneratorRuntime from "@babel/runtime/regenerator";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _objectSpread from "@babel/runtime/helpers/objectSpread";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _applyDecoratedDescriptor from "@babel/runtime/helpers/applyDecoratedDescriptor";

var _dec, _class, _class2, _class3, _temp;

import React, { PureComponent } from 'react';
import { createPortal } from 'react-dom';
import themed from '@rentpath/react-themed';
import autobind from 'autobind-decorator';
import PropTypes from 'prop-types';
import get from 'lodash/fp/get'; //
// Expects currentModal to come in as an object that looks like
// {
//   id: modalId,
//   props: { ...modalProps },
// }
//
// Expects modalDefinitions to come in as an object that looks like
// {
//   modalId: {
//     name: modalName,
//     overlay: isOverlayOpen,
//     resolve: () => import(modalPath),
//   },
// }
//

var ModalStack = (_dec = themed(['Overlay-lock'], {
  pure: true
}), _dec(_class = (_class2 = (_temp = _class3 =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(ModalStack, _PureComponent);

  function ModalStack(props) {
    var _this;

    _classCallCheck(this, ModalStack);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(ModalStack).call(this, props));
    _this.state = {
      currentDefinition: {},
      modals: {}
    };
    return _this;
  }

  _createClass(ModalStack, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.setupWrapperHost();
      this.loadCurrentModal();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      this.loadCurrentModal();
    }
  }, {
    key: "onClose",
    value: function onClose() {
      var _this$props = this.props,
          currentModal = _this$props.currentModal,
          onClose = _this$props.onClose;
      if (onClose) onClose(currentModal.id);
    }
  }, {
    key: "getModalComponent",
    value: function getModalComponent(id) {
      return this.state.modals[id];
    }
  }, {
    key: "setupWrapperHost",
    value: function setupWrapperHost() {
      var modalPortalId = this.props.modalPortalId;
      this.modalHost = document.getElementById(modalPortalId);

      if (!this.modalHost) {
        this.modalHost = document.createElement('section');
        this.modalHost.id = modalPortalId;
        document.body.appendChild(this.modalHost);
      }
    }
  }, {
    key: "loadCurrentModal",
    value: function loadCurrentModal() {
      var _this2 = this;

      var _this$props2 = this.props,
          currentModal = _this$props2.currentModal,
          modalDefinitions = _this$props2.modalDefinitions;
      var id = currentModal && currentModal.id;

      if (id && !this.getModalComponent(id)) {
        var modalDefinition = get(id)(modalDefinitions);
        var modals = this.state.modals;

        if (modalDefinition) {
          this.loadModal(modalDefinition).then(function (modal) {
            if (modal) {
              _this2.setState({
                currentDefinition: modalDefinition,
                modals: _objectSpread({}, modals, _defineProperty({}, id, modal))
              });
            }
          });
        }
      }
    }
  }, {
    key: "loadModal",
    value: function () {
      var _loadModal = _asyncToGenerator(
      /*#__PURE__*/
      _regeneratorRuntime.mark(function _callee(definition) {
        var result;
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return definition.resolve();

              case 2:
                result = _context.sent;
                return _context.abrupt("return", result[definition.name] || result.default);

              case 4:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      return function loadModal(_x) {
        return _loadModal.apply(this, arguments);
      };
    }()
  }, {
    key: "toggleBodyClass",
    value: function toggleBodyClass(toggle) {
      if (this.modalHost) {
        // TODO - does anything still rely on this theme prop? If not remove
        var overlayLockCssClass = this.props.theme['Overlay-lock'];

        if (overlayLockCssClass) {
          document.body.classList.toggle(overlayLockCssClass, toggle);
        }

        document.body.classList.toggle('overflow-hidden', toggle);
      }
    }
  }, {
    key: "renderModals",
    value: function renderModals() {
      var currentModal = this.props.currentModal;
      var currentDefinition = this.state.currentDefinition;
      var ModalComponent = this.getModalComponent(currentModal.id);
      var hasOverlay = currentDefinition.overlay;

      if (ModalComponent) {
        this.toggleBodyClass(true);
        return React.createElement(ModalComponent, _extends({
          key: "modal-".concat(currentModal.id),
          isOpen: true,
          hasOverlay: hasOverlay,
          onClose: this.onClose
        }, currentModal.props));
      }

      return null;
    }
  }, {
    key: "render",
    value: function render() {
      var currentModal = this.props.currentModal;

      if (currentModal && currentModal.id && this.modalHost) {
        return createPortal(this.renderModals(), this.modalHost);
      }

      this.toggleBodyClass(false);
      return null;
    }
  }]);

  return ModalStack;
}(PureComponent), _defineProperty(_class3, "propTypes", {
  currentModal: PropTypes.object,
  modalDefinitions: PropTypes.object,
  onClose: PropTypes.func,
  modalPortalId: PropTypes.string,
  overlay: PropTypes.bool,
  theme: PropTypes.object
}), _defineProperty(_class3, "defaultProps", {
  overlay: false,
  modalPortalId: 'region-modal',
  onClose: function onClose() {},
  theme: {}
}), _temp), (_applyDecoratedDescriptor(_class2.prototype, "onClose", [autobind], Object.getOwnPropertyDescriptor(_class2.prototype, "onClose"), _class2.prototype), _applyDecoratedDescriptor(_class2.prototype, "renderModals", [autobind], Object.getOwnPropertyDescriptor(_class2.prototype, "renderModals"), _class2.prototype)), _class2)) || _class);
export { ModalStack as default };