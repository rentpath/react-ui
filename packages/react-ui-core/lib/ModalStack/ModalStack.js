'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getOwnPropertyDescriptor = require('babel-runtime/core-js/object/get-own-property-descriptor');

var _getOwnPropertyDescriptor2 = _interopRequireDefault(_getOwnPropertyDescriptor);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _extends3 = require('babel-runtime/helpers/extends');

var _extends4 = _interopRequireDefault(_extends3);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _desc, _value, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _autobindDecorator = require('autobind-decorator');

var _autobindDecorator2 = _interopRequireDefault(_autobindDecorator);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _get = require('lodash/fp/get');

var _get2 = _interopRequireDefault(_get);

var _getOr = require('lodash/fp/getOr');

var _getOr2 = _interopRequireDefault(_getOr);

var _Overlay = require('../Modal/Overlay');

var _Overlay2 = _interopRequireDefault(_Overlay);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

//
// Expects currentModal to come in as an object that looks like
// {
//   id: modalId,
//   props: { ...modalProps },
// }
//
// Expects modalDefinitions to come in as an object that looks like
// {
//   modalId: {
//     name: modalName,
//     overlay: isOverlayOpen,
//     resolve: () => import(modalPath),
//   },
// }
//

var ModalStack = (_class = (_temp = _class2 = function (_PureComponent) {
  (0, _inherits3.default)(ModalStack, _PureComponent);

  function ModalStack(props) {
    (0, _classCallCheck3.default)(this, ModalStack);

    var _this = (0, _possibleConstructorReturn3.default)(this, (ModalStack.__proto__ || (0, _getPrototypeOf2.default)(ModalStack)).call(this, props));

    _this.state = {
      currentDefinition: {},
      modals: {}
    };
    return _this;
  }

  (0, _createClass3.default)(ModalStack, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.setupWrapperHost();
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      var _this2 = this;

      // Change to didRecieveProps in future - though, in async rendering update,
      // can resolve the correct component earlier
      var _props = this.props,
          _props$currentModal$i = _props.currentModal.id,
          id = _props$currentModal$i === undefined ? null : _props$currentModal$i,
          modalDefinitions = _props.modalDefinitions;


      if (id && !this.getModalComponent(id)) {
        var modalDefinition = (0, _get2.default)(id)(modalDefinitions);

        if (modalDefinition) {
          this.loadModal(modalDefinition).then(function (modal) {
            _this2.setState({
              currentDefinition: modalDefinition,
              modals: (0, _extends4.default)({}, _this2.state.modals, (0, _defineProperty3.default)({}, id, (0, _getOr2.default)(modal, 'default')(modal)))
            });
          });
        }
      }
    }
  }, {
    key: 'onClose',
    value: function onClose() {
      var _props2 = this.props,
          currentModal = _props2.currentModal,
          onClose = _props2.onClose;


      if (onClose) onClose(currentModal.id);
    }
  }, {
    key: 'getModalComponent',
    value: function getModalComponent(id) {
      return this.state.modals[id];
    }
  }, {
    key: 'setupWrapperHost',
    value: function setupWrapperHost() {
      var modalPortalId = this.props.modalPortalId;


      this.modalHost = document.getElementById(modalPortalId);
      if (!this.modalHost) {
        this.modalHost = document.createElement('section');
        this.modalHost.id = modalPortalId;
        document.body.appendChild(this.modalHost);
      }
    }
  }, {
    key: 'getWrapperComponent',
    value: function getWrapperComponent(useOverlay) {
      var currentModal = this.props.currentModal;

      return useOverlay ? {
        Component: _Overlay2.default,
        props: {
          key: 'overlay-' + currentModal.id,
          onClick: this.onClose
        }
      } : {
        Component: _react.Fragment,
        props: {}
      };
    }
  }, {
    key: 'loadModal',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(definition) {
        var result;
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return definition.resolve();

              case 2:
                result = _context.sent;
                return _context.abrupt('return', result[definition.name]);

              case 4:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function loadModal(_x) {
        return _ref.apply(this, arguments);
      }

      return loadModal;
    }()
  }, {
    key: 'renderModals',
    value: function renderModals() {
      var currentModal = this.props.currentModal;
      var currentDefinition = this.state.currentDefinition;

      var ModalComponent = this.getModalComponent(currentModal.id);

      var Wrapper = this.getWrapperComponent(currentDefinition.overlay);

      if (ModalComponent) {
        return _react2.default.createElement(
          Wrapper.Component,
          Wrapper.props,
          _react2.default.createElement(ModalComponent, (0, _extends4.default)({
            key: 'modal-' + currentModal.id,
            isOpen: true,
            onClose: this.onClose
          }, currentModal.props))
        );
      }
      return null;
    }
  }, {
    key: 'render',
    value: function render() {
      var currentModal = this.props.currentModal;


      if (currentModal && currentModal.id && this.modalHost) {
        return (0, _reactDom.createPortal)(this.renderModals(), this.modalHost);
      }
      return null;
    }
  }]);
  return ModalStack;
}(_react.PureComponent), _class2.propTypes = {
  currentModal: _propTypes2.default.object,
  modalDefinitions: _propTypes2.default.object,
  onClose: _propTypes2.default.func,
  modalPortalId: _propTypes2.default.string
}, _class2.defaultProps = {
  overlay: false,
  modalPortalId: 'region-modal',
  onClose: function onClose() {}
}, _temp), (_applyDecoratedDescriptor(_class.prototype, 'onClose', [_autobindDecorator2.default], (0, _getOwnPropertyDescriptor2.default)(_class.prototype, 'onClose'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'getWrapperComponent', [_autobindDecorator2.default], (0, _getOwnPropertyDescriptor2.default)(_class.prototype, 'getWrapperComponent'), _class.prototype), _applyDecoratedDescriptor(_class.prototype, 'renderModals', [_autobindDecorator2.default], (0, _getOwnPropertyDescriptor2.default)(_class.prototype, 'renderModals'), _class.prototype)), _class);
exports.default = ModalStack;