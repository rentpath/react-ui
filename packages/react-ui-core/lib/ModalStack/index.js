'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _ModalStack = require('./ModalStack');

Object.defineProperty(exports, 'ModalStack', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ModalStack).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }