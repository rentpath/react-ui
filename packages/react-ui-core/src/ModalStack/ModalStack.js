import React, { PureComponent, Fragment } from 'react'
import { createPortal } from 'react-dom'
import autobind from 'autobind-decorator'
import PropTypes from 'prop-types'
import get from 'lodash/fp/get'
import getOr from 'lodash/fp/getOr'
import Overlay from '../Modal/Overlay'

//
// Expects currentModal to come in as an object that looks like
// {
//   id: modalId,
//   props: { ...modalProps },
// }
//
// Expects modalDefinitions to come in as an object that looks like
// {
//   modalId: {
//     name: modalName,
//     overlay: isOverlayOpen,
//     resolve: () => import(modalPath),
//   },
// }
//

export default class ModalStack extends PureComponent {
  static propTypes = {
    currentModal: PropTypes.object,
    modalDefinitions: PropTypes.object,
    onClose: PropTypes.func,
    modalPortalId: PropTypes.string,
  }

  static defaultProps = {
    overlay: false,
    modalPortalId: 'region-modal',
    onClose: () => { },
  }

  constructor(props) {
    super(props)
    this.state = {
      currentDefinition: {},
      modals: {},
    }
  }

  componentDidMount() {
    this.setupWrapperHost()
  }

  componentDidUpdate() {
    // Change to didRecieveProps in future - though, in async rendering update,
    // can resolve the correct component earlier
    const { currentModal: { id = null }, modalDefinitions } = this.props

    if (id && !this.getModalComponent(id)) {
      const modalDefinition = get(id)(modalDefinitions)

      if (modalDefinition) {
        this.loadModal(modalDefinition).then(modal => {
          this.setState({
            currentDefinition: modalDefinition,
            modals: {
              ...this.state.modals,
              // Need to test this, but this should help in case we are in an app
              // that doesn't do auto module export of default
              [id]: getOr(modal, 'default')(modal),
            },
          })
        })
      }
    }
  }

  @autobind
  onClose() {
    const { currentModal, onClose } = this.props

    if (onClose) onClose(currentModal.id)
  }

  getModalComponent(id) {
    return this.state.modals[id]
  }

  setupWrapperHost() {
    const { modalPortalId } = this.props

    this.modalHost = document.getElementById(modalPortalId)
    if (!this.modalHost) {
      this.modalHost = document.createElement('section')
      this.modalHost.id = modalPortalId
      document.body.appendChild(this.modalHost)
    }
  }

  @autobind
  getWrapperComponent(useOverlay) {
    const { currentModal } = this.props
    return useOverlay
      ? {
        Component: Overlay,
        props: {
          key: `overlay-${currentModal.id}`,
          onClick: this.onClose,
        },
      }
      : {
        Component: Fragment,
        props: {},
      }
  }

  async loadModal(definition) {
    const result = await definition.resolve()
    return result[definition.name]
  }

  @autobind
  renderModals() {
    const { currentModal } = this.props
    const { currentDefinition } = this.state
    const ModalComponent = this.getModalComponent(currentModal.id)

    const Wrapper = this.getWrapperComponent(currentDefinition.overlay)

    if (ModalComponent) {
      return (
        <Wrapper.Component {...Wrapper.props}>
          <ModalComponent
            key={`modal-${currentModal.id}`}
            isOpen
            onClose={this.onClose}
            {...currentModal.props}
          />
        </Wrapper.Component>
      )
    }
    return null
  }

  render() {
    const { currentModal } = this.props

    if (currentModal && currentModal.id && this.modalHost) {
      return createPortal(this.renderModals(), this.modalHost)
    }
    return null
  }
}
