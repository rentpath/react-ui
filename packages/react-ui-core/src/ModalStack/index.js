export { default as ModalStack } from './ModalStack'
