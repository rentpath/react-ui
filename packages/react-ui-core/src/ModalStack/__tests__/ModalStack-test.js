import React from 'react'
import renderer from 'react-test-renderer'
import ModalStack from '../ModalStack'

const modalDefinitions = {
  1: {
    name: 'TestModal',
    resolve: () => import('../../LeadModal/LeadModal'),
    overlay: true,
  },
}

const currentModal = {
  id: 1,
  props: {
    listingId: '100057144',
    endecaId: '100057144_1330723',
    propertyLabel: 'LEVEL Furnished Living',
    desktopPhone: '2135684026',
    tplSource: 'MYADS',
    isPaid: true,
    revenue: '17.16',
  },
}

describe('ModalStack', () => {
  it('renders correctly', () => {
    const snap = renderer
      .create(<ModalStack modalDefinitions={modalDefinitions} />)
      .toJSON()
    expect(snap).toMatchSnapshot()
  })
  it('renders a modal correctly', () => {
    const snap = renderer
      .create(<ModalStack currentModal={currentModal} modalDefinitions={modalDefinitions} />)
      .toJSON()
    expect(snap).toMatchSnapshot()
  })
})
