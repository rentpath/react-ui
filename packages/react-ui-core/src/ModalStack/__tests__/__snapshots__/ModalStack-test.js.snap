// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalStack renders a modal correctly 1`] = `null`;

exports[`ModalStack renders correctly 1`] = `null`;
